### DISCUSSION TITLE
Loom and Thread Fairness - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#296] [Announcements]
# Loom and Thread Fairness - Gunnar Morling

Project Loom (JEP 425) is probably amongst the most awaited feature additions to Java ever;
its implementation of virtual threads (or "green threads") promises developers the ability to create highly concurrent applications,
for instance with hundreds of thousands of open HTTP connections,
sticking to the well-known thread-per-request programming model,
without having to resort to less familiar and often more complex to use reactive approaches.

Having been in the workings for several years, Loom got merged into the mainline of OpenJDK just recently and is available as a preview feature in the latest Java 19 early access builds.
I.e. it’s the perfect time to get your hands onto virtual threads and explore the new feature.
In this post I’m going to share an interesting aspect I learned about thread scheduling fairness for CPU-bound workloads running on Loom.

https://www.morling.dev/blog/loom-and-thread-fairness/

**Comment by purplefox** [comment]
As a concrete example of where this could be a problem: 
Imagine a web framework that execute web handlers using Loom virtual threads. Most of the web handlers don't do much CPU processing, but there's one which does - maybe it's doing some crypto stuff, or paring a big xml file which has been uploaded. This processing takes about 5 seconds.
While the virtual threads are busy processing the expensive web handler, no other requests can be handled and the server will hang.

**Reply by safetytrick** [reply]
This only becomes a problem if you have enough of these cpu intensive tasks to occupy the time of all of your carrier threads, and, as mentioned in the article, there are as many carrier threads as there are cores on the system. So if you are allowing enough work in queue to run to take up 100% of your cpu then you would also have a resource problem with traditional threads.

**Reply by purplefox** [reply]
If there were 4 cores on your machine you would only need 4 concurrent uploads of big xml files to hang the system. Easily done.
If you had a traditional OS thread per connection model, it wouldn't have this problem as the OS would timeshare between all threads.
Better still, if you had a virtual thread implementation that timeshared between virtual threads it also wouldn't hang and wouldn't need any more than num_cores carrier threads.

**Reply by nikita2206** [reply]
Just throwing in a simple fix to this: you'd have to make a separate OS thread for this comp-heavy processing, in fact having a separate ExecutorService for heavy tasks in the web context while using green threads for everything else would make things more controllable in the end than if you were using an OS-thread per web request. Because this way you can easily limit how much threads are devoted to the comp-heavy tasks (by configuring the comp-heavy-specific ExecutorService), while letting everything else work on top of green threads.

**Comment by dhchain** [comment]
I am working on a java project https://github.com/dhchain/dhc that uses threads a lot. Of course I am using thread-per-request programming model and have to limit number of connections and requests somehow otherwise the app will become irresponsible or runs out of memory. For replying I did an optimization when I queue responses and let a small pool of threads process them as they come. But for receiving requests I had to keep thread per connection waiting on sockets.
